src/main/kotlin/org/jitsi/nlj/transform/node/incoming/Vp9Parser.kt

```diff
@@ -0,0 +1,144 @@
+        val vp9Packet = packetInfo.packet as Vp9Packet
+
+        ssrcsSeen.add(vp9Packet.ssrc)
+
+        if (vp9Packet.hasScalabilityStructure) {
+            // TODO: handle case where new SS is from a packet older than the
+            //  latest SS we've seen.
+            val packetSpatialLayers = vp9Packet.scalabilityStructureNumSpatial
+            if (packetSpatialLayers != -1) {
+                if (numSpatialLayers != -1 && numSpatialLayers != vp9Packet.scalabilityStructureNumSpatial) {
+                    packetInfo.layeringChanged = true
+                }
+                numSpatialLayers = packetSpatialLayers
+            }
+            val srcAndEnc = findRtpEncodingDesc(vp9Packet)
+            if (srcAndEnc != null) {
+                val (src, enc) = srcAndEnc
+                val newEnc = vp9Packet.getScalabilityStructure(eid = enc.layers[0].eid)
+                if (newEnc != null) {
+                    src.setEncodingLayers(newEnc.layers, vp9Packet.ssrc)
+                }
+                for (otherEnc in src.rtpEncodings) {
+                    if (!ssrcsSeen.contains(otherEnc.primarySSRC)) {
+                        src.setEncodingLayers(emptyArray(), otherEnc.primarySSRC)
+                    }
+                }
+            }
+
+            /* TODO: we need a way to restore the encoding desc's old layer set if it switches back to some other codec
+             *  (i.e. VP8)
+             */
+        }
+        /* VP9 marks keyframes in every packet of the keyframe - only count the start of the frame so the count is correct. */
+        /* Alternately we could keep track of keyframes we've already seen, by timestamp, but that seems unnecessary. */
+        if (vp9Packet.isKeyframe && vp9Packet.isStartOfFrame) {
+            logger.cdebug { "Received a keyframe for ssrc ${vp9Packet.ssrc} ${vp9Packet.sequenceNumber}" }
+            numKeyframes++
+        }
+        if (vp9Packet.spatialLayerIndex > 0 && vp9Packet.isInterPicturePredicted) {
+            /* Check if this layer is using K-SVC. */
+            /* TODO: should we somehow track the keyframe layers separately in
+                the layer descriptions, to get the bitrates more accurate? */
+            findRtpLayerDesc(vp9Packet)?.useSoftDependencies = vp9Packet.usesInterLayerDependency
+        }
+
+        pictureIdState.setState(vp9Packet.hasPictureId, vp9Packet) {
+            "Packet Data: ${vp9Packet.toHex(80)}"
+        }
+        extendedPictureIdState.setState(vp9Packet.hasExtendedPictureId, vp9Packet) {
+            "Packet Data: ${vp9Packet.toHex(80)}"
+        }
+    }
+
+    override fun handleEvent(event: Event) {
+        when (event) {
+            is SetMediaSourcesEvent -> {
+                sources = event.mediaSourceDescs
+            }
+        }
+        super.handleEvent(event)
+    }
+
+    /* TODO: this return value is clumsy */
+    private fun findRtpEncodingDesc(packet: VideoRtpPacket): Pair<MediaSourceDesc, RtpEncodingDesc>? {
```

Comment:
We use this same method name elsewhere to return only the `RtpEncodingDesc`, maybe a better name?